onnx/defs/tensor/defs.cc

```diff
@@ -1185,4 +1185,172 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrains to boolean tensors."));
+
+static const char* OneHot_ver9_doc = R"DOC(
+    Produces a one-hot tensor based on inputs.
+    The locations represented by the index values in the 'indices' input tensor will have the 'on_value' 
+    and the other locations will have the 'off_value' in the output tensor, where 'on_value' and 'off_value' 
+    are specified as part of input argument 'values', which is a two-element tensor of format [off_value, on_value]. 
+    The rank of the output tensor will be one greater than the rank of the input tensor. The additional 
+    dimension is for one-hot representation. The additional dimension will be inserted at the position 
+    specified by 'axis'. If 'axis' is not specified then then additional dimension will be inserted as 
+    the innermost dimension, i.e. axis=-1. The size of the additional dimension is specified by required 
+    scalar input 'depth'. The type of the output tensor can be specified by 'dtype' attribute. If 'dtype' 
+    is not provided, then the type of the output tensor is the same as the type of the 'values' input.
+    Any entries in the 'indices' input tensor with values outside the range [0, depth) will result 
+    in one-hot representation with all 'off_value' values in the output tensor.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    OneHot,
+    9,
+    OpSchema()
+        .SetDoc(OneHot_ver9_doc)
+        .Attr(
+            "axis",
+            "(Optional) Axis along which one-hot representation in added. Default: axis=-1.",
+            AttributeProto::INT,
+            static_cast<int64_t>(-1))
+        .Attr(
+            "dtype",
+            "(Optional) The data type for the elements of the output tensor. If not specified, the type of "
+            "inputs 'values' is used. If neither is provided, the output tensor type defaults to tensor(float32). "
+            "If 'dtype' is string, then input 'values' must be provided.",
+            AttributeProto::INT,
+            OPTIONAL)
+        .Input(
+            0,
+            "indices",
+            "Input tensor containing indices. The values must be non-negative integers. "
+            "Any entries in the 'indices' input tensor with values outside the range [0, depth) "
+            "will result in one-hot representation with all 'off_value' values in the output tensor."
+            "In case 'indices' is of non-integer type, the values will be casted to int64 before use.",
+            "T1")
+        .Input(
+            1,
+            "depth",
+            "Scalar specifying the number of classes in one-hot tensor. This is also the size "
+            "of the one-hot dimension (specified by 'axis' attribute) added on in the output "
+            "tensor and the values in the 'indices' input tensor are expected to be "
+            "in the range [0, depth). The"
+            "In case 'depth' is of non-integer type, it will be casted to int64 before use.",
+            "T1")
+        .Input(
```

Comment:
how about we remove attribute "dtype" and have this input as required? looks to me it's more simple.

Reply:
@linkerzhang - Yes, we could do that. A minor reason to keep it this way could be the situation in packages such as TF, when `values` are available in a different datatype (from prior computation) than the type specified by the explicit `dtype` argument. In such a situation the exporter will have to insert an additional `Cast` node before `OneHot`. Having `dtype` in the ONNX op may provide some convenience and eliminate the need for the extra op.
This is a minor point and I am OK either way. What do you think?

Does anyone else have any preference/feedback?

Reply:
Hmmm, agree this is a minor point. but don't quite understand the case you said. "The type of 'values' should be the same as 'dtype', if 'dtype' is specified." this means anyway, the values's type is indeed indicating output's type, if values are always specified, then output type is also specified (no datatype attribute needed). If values are from previous node ('s output) and the type is not expected, then a CAST is anyway needed, am I right?

Reply:
Yes, you are right, and the `Cast` op is needed anyway. I agree that the `dtype` attribute is not really adding much and just making `values` a required input is a cleaner design. Thanks for pointing this out. Will update the signature.

Reply:
@linkerzhang as per your suggestion, I have removed `dtype` and made `values` a required input.